QUICKSTEP-73: Multi-attribute PartitionSchemeHeader.

zuyu opened the pull request:
This PR added multi-attribute partitions support.

```
CREATE TABLE foo5 (attr1 INT, attr2 LONG, attr3 FLOAT, attr4 DOUBLE, attr5 CHAR(5),
                   attr6 VARCHAR(4))
PARTITION BY HASH(attr1, attr2, attr3, attr4, attr5, attr6) PARTITIONS 4;
```

hbdeshmukh commented:
Hi @zuyu 

Looks like a compilation error is causing travis to fail.

hbdeshmukh commented on catalog/PartitionSchemeHeader.cpp:
```diff
@@ -32,16 +34,29 @@
   DCHECK_GT(num_partitions, 0u);
-  DCHECK_GE(attr_id, 0);
+
+#ifdef QUICKSTEP_DEBUG
```
Optional, but a couple of lines regarding what you are trying to do here should be helpful.

zuyu replied:
Done.

hbdeshmukh commented on catalog/PartitionSchemeHeader.cpp:
```diff
@@ -77,32 +95,44 @@ PartitionSchemeHeader* PartitionSchemeHeader::ReconstructFromProto(
       LOG(FATAL) << "Invalid partition scheme header.";
   }
+
+  return nullptr;
```
Isn't this return statement redundant?

zuyu replied:
In [Travis CI](https://travis-ci.org/apache/incubator-quickstep/jobs/212429450#L1226), the compiler complains `-Werror=return-type` here.

hbdeshmukh commented on catalog/PartitionSchemeHeader.hpp:
```diff
@@ -48,6 +50,13 @@ class Type;
+
+  // A boundary for multi-attribute partitions.
+  // PartitionValues.size() should be equal to PartitionAttributeIds.size().
+  typedef std::vector<TypedValue> PartitionValues;
```
A clearer and more informative name should be ``PartitionBoundaryValues``.

zuyu replied:
Sorry, the comment is not that clear. This represents the corresponding values for `PartitionAttributeIds`, and is also used to store the partition range boundaries for a range-based partition.

hbdeshmukh commented on catalog/PartitionSchemeHeader.hpp:
```diff
@@ -180,16 +190,17 @@ class HashPartitionSchemeHeader : public PartitionSchemeHeader {
    *
-   * @param value_of_attribute The attribute value for which the
-   *                           partition id is to be determined.
+   * @param value_of_attributes The attribute value for which the
```
Nitpick - attribute values instead of attribute value.

hbdeshmukh commented on catalog/PartitionSchemeHeader.hpp:
```diff
@@ -180,16 +190,17 @@ class HashPartitionSchemeHeader : public PartitionSchemeHeader {
    **/
   partition_id getPartitionId(
-      const TypedValue &value_of_attribute) const override {
+      const PartitionValues &value_of_attributes) const override {
```
In the type definition of ``PartitionValues``, you said it is for representing the boundary for partition values. That doesn't seem true here. Why not just use the vector of ``TypedValue`` instead?

hbdeshmukh commented on catalog/PartitionSchemeHeader.hpp:
```diff
@@ -249,20 +269,22 @@ class RangePartitionSchemeHeader : public PartitionSchemeHeader {
    *
-   * @param value_of_attribute The attribute value for which the
-   *                           partition id is to be determined.
+   * @param value_of_attributes The attribute value for which the
```
Nitpick - values instead of value.

hbdeshmukh commented on catalog/PartitionSchemeHeader.hpp:
```diff
@@ -288,22 +310,46 @@ class RangePartitionSchemeHeader : public PartitionSchemeHeader {
+
+    for (std::size_t attr_index = 0; attr_index < partition_attribute_ids_.size(); ++attr_index) {
+      if (less_unchecked_comparators_[attr_index]->compareTypedValues(lhs[attr_index], rhs[attr_index])) {
+        break;
```
Do you mean to use ``continue`` here? The break will cause the control to come out of the for loop and we wouldn't have compared all the attributes, subsequently returning true. Am I missing something?

hbdeshmukh replied:
I think your matching definition is as follows:
(x1, x2, x3, ..., xk) < (y1, y2, y3, ..., yk) if (x1 < x2) || (x2, x3, ..., xk) < (y2, y3, ... yk). 

If that's correct, this should be clearly stated somewhere in the code.

zuyu replied:
We use [lexicographical order](https://en.wikipedia.org/wiki/Lexicographical_order) here, `Given two different sequences of the same length, a1a2...ak and b1b2...bk, the first one is smaller than the second one for the lexicographical order, if ai<bi (for the order of A), for the first i where ai and bi differ.`

hbdeshmukh commented on catalog/tests/PartitionScheme_unittest.cpp:
```diff
@@ -343,15 +363,43 @@ TEST(PartitionSchemeHeaderTest, VarCharRangePartitionSchemeHeaderTest) {
+  vector<PartitionSchemeHeader::PartitionValues> partition_ranges;
+  // Partition boundaries are { 0, 0.00000 }, { 10, 10.0000 }, { 20, 20.00000 }
+  // Last partition can hold upto infinity.
+  // First partition can hold from { -infinity, -infinity } to { 0, -2^(-1074) }.
```
Can you explain -2^(-1074)? I thought the first partition should be < {0, 0.0000} and > {-Inf, -Inf}.

zuyu replied:
Sure, and we are talk about the same thing. In fact, `-2^(-1074)` is the largest negative floating point.

hbdeshmukh commented on catalog/tests/PartitionScheme_unittest.cpp:
```diff
@@ -611,6 +665,7 @@ TEST(PartitionSchemeTest, CheckRangePartitionSchemeSerialization) {
     EXPECT_EQ(blocks_in_part_scheme, blocks_in_part_scheme_from_proto);
   }
 }
+*/
```
Is there some issue with the commented code?

zuyu replied:
Yes, this test triggers [a debug check](https://travis-ci.org/apache/incubator-quickstep/jobs/213492259#L3473) in protobuf in the debug builds only.

hbdeshmukh replied:
Let's not check in commented code. We can either get rid of these tests or try to fix it. I am not sure what these tests are trying to verify. If the functionality being tested is not really critical, we can remove the commented code, create a JIRA issue and move on.

hbdeshmukh commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -724,6 +708,25 @@ const S::PartitionSchemeHeader* Resolver::resolvePartitionClause(
+    if (attr_id == kInvalidAttributeID) {
+      THROW_SQL_ERROR_AT(&partition_attribute_name)
+          << "The given attribute was not found.";
+    } else if (unique_partition_attrs.count(attr_id)) {
```
For readability, can we change this to ``count > 0`` or use ``find`` instead?

hbdeshmukh commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -724,6 +708,25 @@ const S::PartitionSchemeHeader* Resolver::resolvePartitionClause(
     THROW_SQL_ERROR_AT(partition_type_string) << "Unrecognized partition type: " << partition_type;
   }
 
+  proto->set_num_partitions(partition_clause->num_partitions()->long_value());
```
Do we perform error checks for the following conditions?
1. Number of partitions = 0
2. No attribute specified in the partitioning attributes' list.

zuyu replied:
Both cases are covered in the parser, not in the optimizer.

hbdeshmukh commented on storage/InsertDestination.cpp:
```diff
@@ -682,8 +685,11 @@ void PartitionAwareInsertDestination::bulkInsertTuplesWithRemappedAttributes(
+      PartitionSchemeHeader::PartitionValues values;
+      for (const attribute_id attr_id : partition_scheme_header_->getPartitionAttributeIds()) {
+        values.push_back(accessor->getTypedValue(attr_id));
+      }
```
Did you take a look at the function ``getTupleWithAttributes()``. It may be useful here. Moreover the ``tuple`` class can compute hash using the ``getTupleHash`` function.

zuyu replied:
I just checked the methods you mentioned, but I don't think we could use them, as we also need to get partition id in a range partition.

hbdeshmukh commented:
Hi @zuyu 

Finished one round of review. Please let me know when should I take another look.

zuyu commented:
Hi @hbdeshmukh,

Thank you for the review. I have addressed most of issues, and please take another look.

And I also would like to propose to remove the debug builds in Travis CI.

hbdeshmukh commented:
Hi @zuyu 

Looks good to me. I just have one last comment about the commented code in unit tests. After that, we should be good to go. Also, there are two commits for this PR, would you mind squashing them to one?

zuyu commented:
Hi @hbdeshmukh,

I have reported QUICKSTEP-86 for the test failure, and squashed all the commits. Please merge it. Thanks!
